### PR TITLE
feat(mesh): EpochMaxWins merge helper for rate-limit counters

### PR DESCRIPTION
**Pull request by CatherineSue**
### Problem

Step 4 §7.3 (`RateLimitSyncAdapter`) requires an epoch-aware max-wins merge for rate-limit counters. Plain max-wins on the raw count silently undoes window resets: one node resets its counter to 0 at the start of a new window while another still carries the pre-reset count of 100 — plain `max(0, 100) = 100` reverts the reset cluster-wide, defeating the rate limiter.

The `MergeStrategy::EpochMaxWins` variant has sat in `crates/mesh/src/kv.rs` as an enum placeholder since v2 Step 1 with no implementation. Before we can build the adapter that consumes it, we need the merge function itself.

### Solution

Introduce a pure merge helper in `crates/mesh/src/crdt_kv/epoch_max_wins.rs` that:

- Defines the fixed 16-byte wire format: `u64` big-endian epoch in bytes 0–8, `i64` big-endian count in bytes 8–16. Fixed-size + big-endian so both sides of the gossip stream agree byte-for-byte without an application callback.
- Compares **epoch first**, then max-count within the same epoch. A window reset (higher epoch, count=0) always beats a higher count at an older epoch.
- Returns the well-formed side when the other is malformed; keeps `local` when both are malformed. A single corrupt gossip message must never crash the merge loop or replace healthy state with garbage.

This PR only adds the merge function and its unit tests. It does **not** wire `EpochMaxWins` into `CrdtOrMap` or the namespace merge path — that lands together with the `RateLimitSyncAdapter` PR (the fourth in the Step 4 outer sequence). Isolating the merge keeps this reviewable in 200 lines with exhaustive scenario coverage, and lets the adapter PR focus on per-actor sharding and aggregation rather than re-inventing the comparison rule.

## Changes

- **New** `crates/mesh/src/crdt_kv/epoch_max_wins.rs` (~290 lines)
  - `pub const EPOCH_MAX_WINS_ENCODED_LEN: usize = 16`
  - `pub struct EpochCount { epoch: u64, count: i64 }` — owned pair returned from `decode`
  - `pub fn encode(epoch, count) -> [u8; 16]` — stable big-endian layout
  - `pub fn decode(bytes: &[u8]) -> Option<EpochCount>` — `None` on any length ≠ 16
  - `pub fn merge(local: &[u8], remote: &[u8]) -> Vec<u8>` — full merge including malformed fallbacks
  - `#![cfg_attr(not(test), expect(dead_code, reason = "consumed by RateLimitSyncAdapter in a follow-up PR"))]` — gated on non-test so the tests (which exercise every item) don't trip `unfulfilled_lint_expectations`
- `crates/mesh/src/crdt_kv/mod.rs` — registers the new module; no `pub use` yet (the adapter PR will add the re-export when it consumes the helpers).

No public API changes at the crate boundary today. No wire-format changes at the gossip transport layer — this is pure value-level merge logic that the adapter will opt into per-prefix.

## Test Plan

- [x] `cargo test -p smg-mesh --lib crdt_kv::epoch_max_wins` — **12/12 pass** in <1 ms total
- [x] `cargo test -p smg-mesh --lib` — full suite: **256 pass**, 2 ignored (was 248 before this PR; +12 new tests minus the `_ordering_import_used` placeholder removed in Step 3.6 test landing)
- [x] `cargo clippy -p smg-mesh --all-targets --all-features -- -D warnings` clean
- [x] Pre-commit: rustfmt + clippy pass

**Test scenarios** — one test per spec case plus the malformed-input corners and the algebraic properties:

| Test | Scenario |
|------|----------|
| `encode_decode_round_trip` | Six `(epoch, count)` pairs including edge values (`u64::MAX`, `i64::MIN`, negatives) |
| `decode_rejects_wrong_lengths` | 0, 15, 17 bytes all return `None`; exactly 16 is `Some` |
| `same_epoch_max_count_wins` | Normal counting case + commutativity check |
| `higher_epoch_wins_even_with_lower_count` | Window reset propagates: epoch 6 count 0 beats epoch 5 count 30 |
| `lower_epoch_loses_to_local_newer_window` | Stale remote from old window is dropped |
| `near_simultaneous_reset_both_at_zero` | Both sides reset to epoch 5 count 0 → stable 0 |
| `malformed_remote_keeps_local` | Truncated remote does not poison local state |
| `malformed_local_is_replaced_by_remote` | Well-formed remote recovers a corrupted local value |
| `both_malformed_returns_local_no_panic` | Neither side trustworthy → keep local, no panic |
| `signed_count_preserves_sign` | Negative counts preserved across merge (not silently reinterpreted) |
| `merge_is_idempotent` | `merge(v, v) == v` — gossip re-delivery doesn't drift |
| `merge_is_associative_on_three_values` | `((a⊕b)⊕c) == (a⊕(b⊕c))` — required for eventual consistency |

**Not covered in this PR (tracked for follow-ups):**
- Wiring `EpochMaxWins` into `CrdtOrMap::merge` so a configured prefix actually uses this merge rule — goes with the `RateLimitSyncAdapter` PR (fourth in the sequence).
- Per-actor sharding and aggregation (`"rl:{counter}:{node_name}"` key scheme) — also in the adapter PR.
- Rolling-upgrade compat for the v1 per-actor-shard-sum format — spec §2.5 calls out capability-gated migration; that's the adapter PR plus the Step 5 dual-mode-deployment PR.

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes
- [ ] (Optional) Documentation updated
- [ ] (Optional) Please join us on Slack [#sig-smg](https://slack.lightseek.org) to discuss, review, and merge PRs

</details>

## Summary by CodeRabbit

* **Infrastructure**
  * Added an epoch-aware merge for distributed rate-limit counters that deterministically picks the most recent epoch or the highest count when epochs match, improving cross-replica consistency.
  * Built-in validation now rejects malformed or partial counter data, allowing replicas to converge reliably even with corrupted inputs.
  * Preserves signed counter semantics and improves overall reliability of reconciliation under concurrent updates.

### CODE REVIEW

File: crates/mesh/src/crdt_kv/epoch_max_wins.rs [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+//! the caller may choose to encode deltas or reserved sentinels
+//! (e.g. eviction markers) in the same slot; the merge itself only
+//! uses max-compare on the value.
+//!
+//! Malformed input (any length ≠ 16) is kept-if-well-formed-else-
+//! local: a single corrupt gossip message must never crash the merge
+//! loop or retroactively erase a healthy value. If both sides are
+//! malformed, the local copy survives — there's nothing better the
+//! merger can do locally, and a later healthy message from either
+//! side resolves the state.
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/crdt_kv/epoch_max_wins.rs]: Good catch — fixed in e054fdfc. Both-malformed now does a lexicographic byte compare and picks the greater of the two raw slices. Still garbage if both sides are corrupt, but deterministic: `merge(a, b) == merge(b, a)`, so two replicas that arrived at corrupt state from different directions converge instead of diverging indefinitely. A later healthy write from any node restores clean state from the convergence point.

Tests updated: replaced `both_malformed_returns_local_no_panic` with `both_malformed_converges_to_deterministic_winner` (asserts both the winner AND commutativity via the swapped-sides call), plus a new `both_malformed_same_bytes_is_stable` to guard against a `>=` vs `>` regression. Module doc + merge doc comment rewritten to reflect the new behavior.

File: crates/mesh/src/crdt_kv/epoch_max_wins.rs [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+)]
+
+use std::cmp::Ordering;
+
+/// Fixed wire size: 8 bytes big-endian epoch + 8 bytes big-endian count.
+pub const EPOCH_MAX_WINS_ENCODED_LEN: usize = 16;
+
+/// Parsed value from the 16-byte wire format. Returned as an owned
+/// pair rather than a struct to keep this module free of application
+/// types — adapters layer their own newtype on top.
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/crdt_kv/epoch_max_wins.rs]: Fixed in e054fdfc. Rewrote the doc to describe the actual distinction — the decoder returns an owned `EpochCount` so callers don't have to keep the source slice alive across the merge, and the struct stays application-agnostic so adapters (e.g. `RateLimitSyncAdapter`) layer their own newtype on top. Clearer than the previous misleading "pair rather than a struct" wording.

File: crates/mesh/src/crdt_kv/epoch_max_wins.rs [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+            };
+            encode(winner.epoch, winner.count).to_vec()
+        }
+        (Some(_), None) => local.to_vec(),
+        (None, Some(_)) => remote.to_vec(),
+        (None, None) => {
+            if local >= remote {
+                local.to_vec()
+            } else {
+                remote.to_vec()
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/crdt_kv/epoch_max_wins.rs]: Right call — reverted in 322ceb97. The `MergeStrategy::EpochMaxWins` doc in `kv.rs` and the mesh-v2 spec §2.5 both commit to "keeps local" for the malformed/malformed case. I'd briefly switched to byte-lex to address CodeRabbit's commutativity concern, but that contradicted the published contract and broke the no-op-on-store property downstream callers may rely on.

Back to keep-local. The module doc now notes the non-commutativity in the malformed/malformed arm explicitly and explains why it's acceptable given the rate-limit workload (every increment writes, every window resets — a well-formed write restores clean state before non-convergence can matter).

File: crates/mesh/src/crdt_kv/epoch_max_wins.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/crdt_kv/epoch_max_wins.rs]: Switching back on this one. Codex (`r3120750985`) pointed out that the byte-lex fallback violates the published `MergeStrategy::EpochMaxWins` contract in `kv.rs` and mesh-v2 spec §2.5, both of which explicitly say "If both are malformed, keeps local." That contract matters to downstream callers — the storage layer can skip the write when `merged == current`, which is a true no-op that byte-lex breaks.

Reverted to keep-local in 322ceb97. The commutativity concern you raised is real but bounded: the only scenario where non-convergence persists is two replicas both holding corrupt state AND no subsequent writes. Rate-limit counters write on every increment and reset every window, so the first healthy write from anyone restores clean state cluster-wide. The module doc now makes this trade-off explicit so the choice isn't mysterious to future readers.
